### PR TITLE
Outbound ssh

### PR DESCRIPTION
**Pull request by mwiget**
This pull request brings outbound ssh support to PyEZ and depends on the recently  merged outbound-ssh support in ncclient via pull request https://github.com/ncclient/ncclient/pull/196 

Unit test has been added to cover the new functionality. 

A functional test example is provided in tests/functional/test_outbound_ssh.py

### CODE REVIEW

**Comment by jnpr-community-netdev** [comment]
Can one of the admins verify this patch?

**Comment by stacywsmith** [comment]
@mwiget Since this depends on ncclient/ncclient#196, I'd like to wait until ncclient/ncclient#196 is available in a released version of ncclient before we merge this. That will allow us to bump the required version of ncclient in requirements.txt and avoid users experiencing incompatibilities between PyEZ and ncclient. I'll try to review the this pull request in detail soon, but wanted you to have a heads up on why we are waiting to merge.

File: lib/jnpr/junos/device.py [hunk 1]
```diff
@@ -1075,10 +1080,11 @@ def __init__(self, *vargs, **kvargs):
             self._ssh_config = None
         else:
             # --------------------------
-            # making a remote connection
+            # making a remote connection 
+            # or wait for incoming 'outbound-ssh' connection if hostname is None
```
Comment (stacywsmith) [inline lib/jnpr/junos/device.py]: Suggest changing this comment to:
```
# if hostname is None, this is an 'outbound-ssh' connection
# which uses the established TCP connection from sock_fd
```

**Comment by mwiget** [comment]
@stacywsmith thanks for the review. I updated the outbound-ssh comment as requested.

**Comment by jnpr-community-netdev** [comment]
Build finished. 731 tests run, 0 skipped, 0 failed.

**Comment by jeremyschulman** [comment]
@stacywsmith @vnitinv - I will leave this to you to perform the merge.  I have been also waiting for this enhancement.  I would greatly appreciate it if this was part of the upcoming 2.2.0 release.  Thank you!

**Comment by mwiget** [comment]
Example code:

```Python
import socket
from jnpr.junos import Device
from pprint import pprint

"""
  Listen on TCP port 2200 for incoming SSH session from Junos devices with
  the following ssh outbound configuration and collect the devices facts
  upon connect, then disconnect from that device and wait for more connections

  lab@router> show configuration system services outbound-ssh
  client outbound-ssh {
    device-id router1;
    services netconf;
    138.201.122.182 port 2200;
  }
    
"""

def launch_junos_proxy(client, addr):
    val = {
            'MSG-ID': None,
            'MSG-VER': None,
            'DEVICE-ID': None
            }
    msg = ''
    count = 3
    while len(msg) < 100 and count > 0:
        c = client.recv(1)
        if c == '\r':
            continue

        if c == '\n':
            count -= 1
            if msg.find(':'):
                (key, value) = msg.split(': ')
                val[key] = value
                msg = ''
        else:
            msg += c

    print('MSG %s %s %s' %
            (val['MSG-ID'], val['MSG-VER'], val['DEVICE-ID']))
    return client.fileno()

def main():
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    port = 2200
    s.bind(('', port))
    s.listen(5)
    print('\nListening on port %d for incoming sessions ...' % (port))
    sock_fd = 0
    while True:
        client, addr = s.accept()
        print('\nGot a connection from %s:%d!' % (addr[0], addr[1]))
        sock_fd = launch_junos_proxy(client, addr)
        print('Logging in ...')

        dev = Device(host=None, sock_fd=sock_fd)
        dev.open()
#        print('Hostname is %s' % (dev.facts['hostname']))
        pprint(dev.facts)
        dev.close()

if __name__ == "__main__":
    main()
```

Executing the script while 2 Junos devices are configured with outbound-ssh as shown in the script and allowing ssh access via public ssh key by the user running the script provides the following output:

```
$ python outbound_ssh.py 

Listening on port 2200 for incoming sessions ...

Got a connection from 192.168.99.3:60110!
MSG DEVICE-CONN-INFO V1 router1
Logging in ...
{'2RE': False,
 'HOME': '/var/home/mwiget',
 'RE0': {'last_reboot_reason': 'Router rebooted after a normal shutdown.',
         'mastership_state': 'master',
         'model': 'RE-VMX',
         'status': 'OK',
         'up_time': '2 hours, 27 minutes, 16 seconds'},
 'RE1': None,
 'RE_hw_mi': False,
 'current_re': ['re0', 'master', 'node', 'fwdd', 'member', 'pfem'],
 'domain': None,
 'fqdn': 'vmx2',
 'hostname': 'vmx2',
 'hostname_info': {'re0': 'vmx2'},
 'ifd_style': 'CLASSIC',
 'junos_info': {'re0': {'object': junos.version_info(major=(17, 3), type=R, minor=2, build=10),
                        'text': '17.3R2.10'}},
 'master': 'RE0',
 'model': 'VMX',
 'model_info': {'re0': 'VMX'},
 'personality': 'MX',
 're_info': {'default': {'0': {'last_reboot_reason': 'Router rebooted after a normal shutdown.',
                               'mastership_state': 'master',
                               'model': 'RE-VMX',
                               'status': 'OK'},
                         'default': {'last_reboot_reason': 'Router rebooted after a normal shutdown.',
                                     'mastership_state': 'master',
                                     'model': 'RE-VMX',
                                     'status': 'OK'}}},
 're_master': {'default': '0'},
 'serialnumber': 'VM5B7BCDF4C6',
 'srx_cluster': None,
 'srx_cluster_id': None,
 'srx_cluster_redundancy_group': None,
 'switch_style': 'BRIDGE_DOMAIN',
 'vc_capable': False,
 'vc_fabric': None,
 'vc_master': None,
 'vc_mode': None,
 'version': '17.3R2.10',
 'version_RE0': '17.3R2.10',
 'version_RE1': None,
 'version_info': junos.version_info(major=(17, 3), type=R, minor=2, build=10),
 'virtual': True}

Got a connection from 192.168.99.4:50713!
MSG DEVICE-CONN-INFO V1 router1
Logging in ...
{'2RE': False,
 'HOME': '/var/home/mwiget',
 'RE0': {'last_reboot_reason': 'Router rebooted after a normal shutdown.',
         'mastership_state': 'master',
         'model': 'RE-VMX',
         'status': 'OK',
         'up_time': '2 hours, 20 minutes, 32 seconds'},
 'RE1': None,
 'RE_hw_mi': False,
 'current_re': ['re0', 'master', 'node', 'fwdd', 'member', 'pfem'],
 'domain': None,
 'fqdn': 'vmx1',
 'hostname': 'vmx1',
 'hostname_info': {'re0': 'vmx1'},
 'ifd_style': 'CLASSIC',
 'junos_info': {'re0': {'object': junos.version_info(major=(17, 3), type=R, minor=2, build=10),
                        'text': '17.3R2.10'}},
 'master': 'RE0',
 'model': 'VMX',
 'model_info': {'re0': 'VMX'},
 'personality': 'MX',
 're_info': {'default': {'0': {'last_reboot_reason': 'Router rebooted after a normal shutdown.',
                               'mastership_state': 'master',
                               'model': 'RE-VMX',
                               'status': 'OK'},
                         'default': {'last_reboot_reason': 'Router rebooted after a normal shutdown.',
                                     'mastership_state': 'master',
                                     'model': 'RE-VMX',
                                     'status': 'OK'}}},
 're_master': {'default': '0'},
 'serialnumber': 'VM5B7BCDF7F5',
 'srx_cluster': None,
 'srx_cluster_id': None,
 'srx_cluster_redundancy_group': None,
 'switch_style': 'BRIDGE_DOMAIN',
 'vc_capable': False,
 'vc_fabric': None,
 'vc_master': None,
 'vc_mode': None,
 'version': '17.3R2.10',
 'version_RE0': '17.3R2.10',
 'version_RE1': None,
 'version_info': junos.version_info(major=(17, 3), type=R, minor=2, build=10),
 'virtual': True}
^CTraceback (most recent call last):
  File "outbound_ssh.py", line 68, in <module>
    main()
  File "outbound_ssh.py", line 54, in main
    client, addr = s.accept()
  File "/usr/lib/python2.7/socket.py", line 206, in accept
    sock, addr = self._sock.accept()
KeyboardInterrupt
```

**Comment by vnitinv** [comment]
Nice tutorial on how to manage ssh keys
https://www.linode.com/docs/security/authentication/use-public-key-authentication-with-ssh/

**Comment by vnitinv** [comment]
We can also pass user credentials in place of ssh keys. For example

```python
    while True:
        client, addr = s.accept()
        print('\nGot a connection from %s:%d!' % (addr[0], addr[1]))
        sock_fd = launch_junos_proxy(client, addr)
        print('Logging in ...')
        print('\n sockfd after %d'%(sock_fd))
        dev = Device(sock_fd=sock_fd, user='xxxx', passwd='xxxx')   <<<<
        dev.open()
```

**Pull request merged**